src/libraries/System.Private.CoreLib/src/System/Globalization/Normalization.Icu.cs

```diff
@@ -114,25 +117,95 @@ private static unsafe string IcuNormalize(string strInput, NormalizationForm nor
             }
         }
 
-        private static void ValidateArguments(string strInput, NormalizationForm normalizationForm)
+        private static unsafe bool IcuTryNormalize(ReadOnlySpan<char> source, Span<char> destination, out int charsWritten, NormalizationForm normalizationForm = NormalizationForm.FormC)
         {
-            Debug.Assert(strInput != null);
+            Debug.Assert(!GlobalizationMode.Invariant);
+            Debug.Assert(!GlobalizationMode.UseNls);
+            Debug.Assert(!source.IsEmpty);
+            Debug.Assert(normalizationForm == NormalizationForm.FormC || normalizationForm == NormalizationForm.FormD || normalizationForm == NormalizationForm.FormKC || normalizationForm == NormalizationForm.FormKD);
 
-            if ((OperatingSystem.IsBrowser() || OperatingSystem.IsWasi())&& (normalizationForm == NormalizationForm.FormKC || normalizationForm == NormalizationForm.FormKD))
+            if (destination.IsEmpty)
             {
-                // Browser's ICU doesn't contain data needed for FormKC and FormKD
-                throw new PlatformNotSupportedException();
+                charsWritten = 0;
+                return false;
+            }
+
+            ValidateArguments(source, normalizationForm, nameof(source));
+
+            int realLen;
+            fixed (char* pInput = source)
+            fixed (char* pDest = destination)
+            {
+#if TARGET_MACCATALYST || TARGET_IOS || TARGET_TVOS
+                if (GlobalizationMode.Hybrid)
+                {
+                    realLen = Interop.Globalization.NormalizeStringNative(normalizationForm, pInput, source.Length, pDest, destination.Length);
+                }
+                else
+#endif
+                {
+                    realLen = Interop.Globalization.NormalizeString(normalizationForm, pInput, source.Length, pDest, destination.Length);
+                }
+            }
+
+            if (realLen == -1)
```

Comment:
```suggestion
            if (realLen < 0)
```
for a tad better machine code.